## Задание №1. Создание и изменение массивов.

Создайте массив NumPy размером 3×3, содержащий числа от 1 до 9. Умножьте все элементы массива на 2, а затем замените все элементы больше 10 на 0. Выведите итоговый массив. 

In [6]:
import numpy as np
arr = np.arange(1, 10).reshape(3, 3)
arr *= 2
arr[arr > 10] = 0

print(arr)

[[ 2  4  6]
 [ 8 10  0]
 [ 0  0  0]]


## Задание №2. Работа с булевыми масками.

Создайте массив NumPy из 20 случайных целых чисел от 1 до 100. Найдите и выведите все элементы, которые делятся на 5 без остатка. Затем замените их на -1 и выведите обновленный массив.

In [24]:
arr = np.random.randint(1, 101, size=20)

mask = arr % 5 == 0
div_by_5 = arr[mask]

arr[mask] = -1

arr, div_by_5

(array([84, 36, 51, 46, 48, 87, -1, 73, -1, 27, 82, 89, 48, 52, 14, 56, 96,
        -1, 32, 47]),
 array([30, 30, 50]))

## Задание №3. Объединение и разбиение массивов.

Создайте два массива NumPy размером 1×5, заполненные случайными числами от 0 до 50.

1. Объедините эти массивы в один двумерный массив (по строкам).
2. Разделите полученный массив на два массива, каждый из которых содержит 5 элементов.

Выведите все промежуточные и итоговые результаты.


In [26]:
arr1 = np.random.randint(0, 51, size=(1, 5))
arr2 = np.random.randint(0, 51, size=(1, 5))

print(arr1, arr2, "\n")

com_arr = np.vstack((arr1, arr2))

print(com_arr, "\n")

split_arr1, split_arr2 = np.vsplit(com_arr, 2)
print("Разделенный массив 1:", split_arr1)
print("Разделенный массив 2:", split_arr2)

[[11 25 12  1  4]] [[41 24 37 45 27]] 

[[11 25 12  1  4]
 [41 24 37 45 27]] 

Разделенный массив 1: [[11 25 12  1  4]]
Разделенный массив 2: [[41 24 37 45 27]]


## Задание №4. Генерация и работа с линейными последовательностями.

Создайте массив из 50 чисел, равномерно распределенных от -10 до 10. Вычислите сумму всех элементов, сумму положительных элементов и сумму отрицательных элементов. Выведите результаты.

In [10]:
arr = np.sort(np.random.uniform(-10, 10, 50))

sum_el = np.sum(arr)
sum_positive = np.sum(arr[arr > 0])
sum_negative = np.sum(arr[arr < 0])

print("Массив:", arr)
print("Сумма всех элементов:", sum_el)
print("Сумма положительных элементов:", sum_positive)
print("Сумма отрицательных элементов:", sum_negative)

Массив: [-8.55623132 -8.42583131 -8.35798474 -7.80489569 -7.51603942 -7.49587102
 -7.18492786 -7.15969978 -7.08321922 -6.57880148 -4.39934585 -4.18405856
 -3.45376981 -3.18501821 -2.90796953 -2.72023315 -2.57631757 -1.61323024
 -1.48714164 -1.25414251 -1.09693359 -0.7578166  -0.52301017  0.13308036
  0.39310265  0.39584567  0.82107651  1.9341879   2.09413437  2.97774192
  3.65411242  4.42165257  4.92959182  5.11711563  5.53610639  6.34953116
  6.49314827  6.62485197  6.69337762  6.77293363  7.00800425  7.52203747
  7.99735444  8.10285982  8.88442494  8.9064871   9.56552001  9.60077134
  9.78618374  9.82191128]
Сумма всех элементов: 46.2146559899438
Сумма положительных элементов: 152.5371452648804
Сумма отрицательных элементов: -106.32248927493656


## Задание №5. Работа  с диагональными и единичными матрицами.

Создайте:
1. Единичную матрицу размером 4×4. 
2. Диагональную матрицу размером 4×4 с диагональными элементами [5, 10, 15, 20] (не использовать циклы).

Найдите сумму всех элементов каждой из этих матриц и сравните результаты.


In [85]:
arr_eye = np.eye(4)
arr_eye_dig = np.diag([5, 10, 15, 20])

sum_arr_eye = np.sum(arr_eye)
sum_arr_eye_dig = np.sum(arr_eye_dig)

print("Единичная матрица:\n", arr_eye)
print("Сумма элементов единичной матрицы:", sum_arr_eye)

print("\nДиагональная матрица:\n", arr_eye_dig)
print("Сумма элементов диагональной матрицы:", sum_arr_eye_dig)

Единичная матрица:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
Сумма элементов единичной матрицы: 4.0

Диагональная матрица:
 [[ 5  0  0  0]
 [ 0 10  0  0]
 [ 0  0 15  0]
 [ 0  0  0 20]]
Сумма элементов диагональной матрицы: 50


## Задание №6. Создание и базовые операции с матрицами.

Создайте две квадратные матрицы NumPy размером 3×3, заполненные случайными целыми числами от 1 до 20. 

Вычислите и выведите:
1. Их сумму
2. Их разность 
3. Их поэлементное произведение



In [88]:
matrix1 = np.matrix(np.random.randint(1, 21, (3, 3)))
matrix2 = np.matrix(np.random.randint(1, 21, (3, 3)))

sum_matrix = matrix1 + matrix2
diff_matrix = matrix1 - matrix2
prod_matrix = np.multiply(matrix1, matrix2) 

print("Первая матрица:\n", matrix1)
print("Вторая матрица:\n", matrix2)
print("Сумма матриц:\n", sum_matrix)
print("Разность матриц:\n", diff_matrix)
print("Поэлементное произведение матриц:\n", prod_matrix)


Первая матрица:
 [[17  2 19]
 [16 17 10]
 [ 1 11 11]]
Вторая матрица:
 [[20  7  9]
 [16  4 15]
 [11 15 11]]
Сумма матриц:
 [[37  9 28]
 [32 21 25]
 [12 26 22]]
Разность матриц:
 [[ -3  -5  10]
 [  0  13  -5]
 [-10  -4   0]]
Поэлементное произведение матриц:
 [[340  14 171]
 [256  68 150]
 [ 11 165 121]]


## Задание №7. Умножение матриц.

Создайте две матрицы NumPy: 

1. Первую размером 2×3, заполненную случайными числами от 1 до 10. 
2. Вторую размером 3×2, заполненную случайными числами от 1 до 10.

Выполните матричное умножение (@ или np.dot) и выведите результат


In [91]:
matrix1 = np.matrix(np.random.randint(1, 11, (2, 3)))
matrix2 = np.matrix(np.random.randint(1, 11, (3, 2)))

dot_matrix = matrix1 @ matrix2

print("Первая матрица:\n", matrix1)
print("Вторая матрица:\n", matrix2)
print("Результат матричного умножения:\n", dot_matrix)

Первая матрица:
 [[10  8 10]
 [ 8  2  8]]
Вторая матрица:
 [[ 8  7]
 [ 8  9]
 [ 1 10]]
Результат матричного умножения:
 [[154 242]
 [ 88 154]]


## Задание №8. Определитель и обратная матрица.

Создайте случайную квадратную матрицу 3×3. Найдите и выведите: 

1. Определитель этой матрицы 
2. Обратную матрицу (если существует, иначе выведите сообщение, что матрица вырождена)
   
Используйте функции np.linalg.det и np.linalg.inv.


In [94]:
matrix = np.matrix(np.random.randint(1, 11, (3, 3)))
determ = np.linalg.det(matrix)

if np.isclose(determ, 0):
    print("Матрица вырожденна")
else:
    print("Обратная матрица:\n", np.linalg.inv(matrix))

Обратная матрица:
 [[ 1.33333333 -2.27777778 -0.11111111]
 [-0.66666667  1.05555556  0.22222222]
 [-1.          2.          0.        ]]


## Задание №9. Транспонирование и след матрицы.

Создайте матрицу NumPy размером 4×4, содержащую случайные целые числа от 1 до 50. Выведите: 
1. Исходную матрицу 
2. Транспонированную матрицу 
3. След матрицы (сумму элементов на главной диагонали)
   
Используйте np.trace для нахождения следа.


In [97]:
matrix = np.matrix(np.random.randint(1, 51, (4, 4)))

transpons_matrix = matrix.T

trace_matrix = np.trace(matrix)

print("Исходная матрица:\n", matrix)
print("Транспонированная матрица:\n", transpons_matrix)
print("След матрицы:", trace_matrix)

Исходная матрица:
 [[43 14 23  6]
 [10 38 33 17]
 [21 41 32 27]
 [40 30  1 42]]
Транспонированная матрица:
 [[43 10 21 40]
 [14 38 41 30]
 [23 33 32  1]
 [ 6 17 27 42]]
След матрицы: 155


## Задание №10. Системы линейных уравнений.

Решить систему линейных уравнений.

\begin{cases}
2x + 3y - z = 5 \\
4x - y + 2z = 6 \\
-3x + 5y + 4z = -2
\end{cases}

Используйте матричное представление Ax = B , где A – матрица коэффициентов,
x – вектор неизвестных, B – вектор правой части. Решите систему с помощью
np.linalg.solve и выведите результат.

In [100]:
A = np.array([
    [2, 3, -1],
    [4, -1, 2],
    [-3, 5, 4]
])
B = np.array([5, 6, -2])

x = np.linalg.solve(A, B)

print("Решение системы:\n", x)

Решение системы:
 [1.63963964 0.57657658 0.00900901]


## Задание №11. Индивидуальное задание.

**Вариант №8.** Перераспределение бюджета. Городской бюджет распределяется между
образованием, здравоохранением и транспортом. На здравоохранение
выделяется в 1.5 раза больше, чем на транспорт, а на образование — на 20%
больше, чем на здравоохранение. Всего в бюджете 5 000 000 рублей. Сколько
денег выделено на каждую сферу?

X - транспорт; Y - здравоохранение; Z - образование


y =  1.5x

z = 1.2y

x + y + z = 5000000

*Система уравнений:*

\begin{cases} 
x + y + z = 5 000 000 \\
-1.5x + y = 0 \\ 
-1.2y + z = 0 
\end{cases}


In [104]:
A = np.array([
    [1, 1, 1,],
    [-1.5, 1, 0],
    [0, -1.2, 1]
])

B = np.array([5000000, 0, 0])

det_A = np.linalg.det(A)

A_x = A.copy()
A_x[:, 0] = B
det_A_x = np.linalg.det(A_x)

A_y = A.copy()
A_y[:, 1] = B
det_A_y = np.linalg.det(A_y)

A_z = A.copy()
A_z[:, 2] = B
det_A_z = np.linalg.det(A_z)

x = det_A_x / det_A
y = det_A_y / det_A
z = det_A_z / det_A

print("Метод крамера:\n")
print(f"x = {x:.2f} рублей")
print(f"y = {y:.2f} рублей")
print(f"z = {z:.2f} рублей")

Метод крамера:

x = 1162790.70 рублей
y = 1744186.05 рублей
z = 2093023.26 рублей


In [106]:
results = np.linalg.inv(A) @ B

print("Решение матричным методом:\n")
print(f"x: {results[0]:.2f} рублей")
print(f"y: {results[1]:.2f} рублей")
print(f"z: {results[2]:.2f} рублей")

Решение матричным методом:

x: 1162790.70 рублей
y: 1744186.05 рублей
z: 2093023.26 рублей


In [108]:
results_solve = np.linalg.solve(A, B)

print("Решение с помощью np.linalg.solve:\n")
print(f"x: {results_solve[0]:.2f} рублей")
print(f"y: {results_solve[1]:.2f} рублей")
print(f"z: {results_solve[2]:.2f} рублей")

Решение с помощью np.linalg.solve:

x: 1162790.70 рублей
y: 1744186.05 рублей
z: 2093023.26 рублей


**Все методы дали одинаковый результат, т.е. ответ:**

Транспорт: 1162790.70 руб.

Здравоохранение: 1744186.05 руб.

Образование: 2093023.26 руб.